### PR TITLE
Add the fixed number type (OSSL_FN) and its allocators and deallocator

### PR DESCRIPTION
**Pull request by levitte**
This includes a small test program that performs introspection of the
OSSL_FN, to check that diverse functions do what's expected of them.

Related-to: doc/designs/fixed-size-large-numbers.md
Resolves: https://github.com/openssl/project/issues/1649

### CODE REVIEW

**Comment by levitte** [comment]
This is currently rebased of top of #28787, to allow the pre-commit workflow to work smoothly

File: crypto/fn/fn_lib.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#include <stdbool.h>
+#include <openssl/opensslconf.h>
+#include <openssl/crypto.h>
+#include "fn_local.h"
+
+#define SIZE_IS_BITS 0
+
+static OSSL_FN *ossl_fn_new_internal(size_t size, bool securely)
+{
+#if SIZE_IS_BITS
```
Comment (mattcaswell) [inline crypto/fn/fn_lib.c]: Not sure. I would tend towards input size in bytes

File: crypto/fn/fn_lib.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     * Open question: should the input size be in bits?  In that case,
+     * this adjustment is necessary.
+     */
+
+    /* Bits to bytes, rounded up */
+    size = (size + CHAR_BITS - 1) / CHAR_BITS;
+#endif
+
+    /* bytes to number of limbs, rounded up */
+    size_t limbs = (size + sizeof(OSSL_FN_ULONG) - 1) / sizeof(OSSL_FN_ULONG);
```
Comment (mattcaswell) [inline crypto/fn/fn_lib.c]: I think we still require no mixed declarations and code??? Although irrelevant if the ifdef above goes.

File: crypto/fn/fn_lib.c [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    return ret;
+}
+
+static void ossl_fn_free_internal(OSSL_FN *f, bool clear)
+{
+    if (f == NULL)
+        return;
+
+    size_t limbssize = f->dsize * sizeof(OSSL_FN_ULONG);
+    size_t totalsize = limbssize + sizeof(OSSL_FN);
```
Comment (mattcaswell) [inline crypto/fn/fn_lib.c:59]: Mixed declarations/code

File: test/fn_internal_test.c [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        || !TEST_size_t_eq(f->dsize, 2)
+        || !TEST_size_t_eq(f->d[0], 0)
+        || !TEST_size_t_eq(f->d[1], 0))
+        ret = 0;
+    if (f != NULL)
+        OSSL_FN_free(f);
+
+    return ret;
+}
+static int test_secure_alloc(void)
```
Comment (mattcaswell) [inline test/fn_internal_test.c:63]: Blank line between function declarations?

File: test/fn_internal_test.c [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    if (!TEST_ptr(f = OSSL_FN_new(sizeof(OSSL_FN_ULONG) * 2 - 2))
+        || !TEST_uint_eq(f->is_dynamically_allocated, 1)
+        || !TEST_uint_eq(f->is_securely_allocated, 0)
+        || !TEST_uint_eq(f->is_acquired, 0)
+        || !TEST_uint_eq(f->is_negative, 0)
+        || !TEST_size_t_eq(f->dsize, 2)
+        || !TEST_size_t_eq(f->d[0], 0)
+        || !TEST_size_t_eq(f->d[1], 0))
+        ret = 0;
+    if (f != NULL)
```
Comment (mattcaswell) [inline test/fn_internal_test.c]: Shouldn't be necessary?

File: test/fn_internal_test.c [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    if (!TEST_ptr(f = OSSL_FN_secure_new(sizeof(OSSL_FN_ULONG) * 2 - 2))
+        || !TEST_uint_eq(f->is_dynamically_allocated, 1)
+        || !TEST_uint_eq(f->is_securely_allocated, 1)
+        || !TEST_uint_eq(f->is_acquired, 0)
+        || !TEST_uint_eq(f->is_negative, 0)
+        || !TEST_size_t_eq(f->dsize, 2)
+        || !TEST_size_t_eq(f->d[0], 0)
+        || !TEST_size_t_eq(f->d[1], 0))
+        ret = 0;
+    if (f != NULL)
```
Comment (mattcaswell) [inline test/fn_internal_test.c]: Shouldn't be necessary?

File: test/fn_internal_test.c [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+int setup_tests(void)
+{
+    ADD_TEST(test_alloc);
+    ADD_TEST(test_secure_alloc);
+
+    return 1;
+}
+
+void cleanup_tests(void)
```
Comment (mattcaswell) [inline test/fn_internal_test.c]: Can't we just omit this?

File: crypto/fn/fn_lib.c (hunk 3 above)
Comment (levitte) [reply crypto/fn/fn_lib.c:59]: The C90 arlternative is this:

``` C
    size_t limbssize;
    size_t totalsize;

    if (f == NULL)
        return;

    limbssize = f->dsize * sizeof(OSSL_FN_ULONG);
    totalsize = limbssize + sizeof(OSSL_FN);
```

My bikeshed says this is uglier and slightly less readable.  ...  and we're not constrained to C90 any longer.

**Comment by levitte** [comment]
I've converted this to a draft for now, 'cause I've discovered that there's code that depends on `dmax` and `top` in BIGNUM being `int`.  I've tested changing them to `size_t`, and hell broke loose when testing.  This is most probably a fairly low level dependence, perhaps even at assembler level.  `sizeof(size_t)` sometimes (often?) differing from `sizeof(int)` seems to be a factor.

That will affect future transitions of BIGNUMs to OSSL_FNs, so I'm taking some time to rethink the OSSL_FN structure.

File: test/fn_internal_test.c (hunk 5 above)
Comment (levitte) [reply test/fn_internal_test.c]: Good point

File: crypto/fn/fn_lib.c (hunk 1 above)
Comment (t8m) [reply crypto/fn/fn_lib.c]: The cryptosystems are usually defined in number of bits not bytes, but perhaps it does not matter much.

File: crypto/fn/fn_lib.c (hunk 2 above)
Comment (t8m) [reply crypto/fn/fn_lib.c]: > I think we still require no mixed declarations and code???

I think some already slipped in.

File: crypto/fn/fn_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/fn/fn_lib.c]: *Somewhere*, we will convert bits to bytes.  The question is where it is most practical.

**Comment by t8m** [comment]
> I've converted this to a draft for now, 'cause I've discovered that there's code that depends on `dmax` and `top` in BIGNUM being `int`. I've tested changing them to `size_t`, and hell broke loose when testing. This is most probably a fairly low level dependence, perhaps even at assembler level. `sizeof(size_t)` sometimes (often?) differing from `sizeof(int)` seems to be a factor.
> 
> That will affect future transitions of BIGNUMs to OSSL_FNs, so I'm taking some time to rethink the OSSL_FN structure.

We can keep them being int (or change to uint32_t ?) There is no point in needing 64bit numbers for size of fixnums.

File: crypto/fn/fn_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/fn/fn_lib.c]: So, who wants to flip a coin and decide?

Comment (mattcaswell) [reply crypto/fn/fn_lib.c]: I say bytes.

**Comment by levitte** [comment]
I've made the change so `dsize` is now an "int".  There's a comment
with a brief note on this, and the commit message includes more of my
reasoning around this.

File: crypto/fn/fn_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/fn/fn_lib.c]: I actually changed that size to limbs, 'cause that matches the BIGNUM internals, and then added allocators for limbs, bytes and bits, so we cover all use cases.

File: crypto/fn/fn_lib.c [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+#include <openssl/opensslconf.h>
+#include <openssl/crypto.h>
+#include "fn_local.h"
+
+#define SIZE_IS_BITS 0
+
+static OSSL_FN *ossl_fn_new_internal(int limbs, bool securely)
+{
+    /* Total size of the whole OSSL_FN, in bytes */
+    int totalsize = sizeof(OSSL_FN) + limbs * sizeof(OSSL_FN_ULONG);
```
Comment (mattcaswell) [inline crypto/fn/fn_lib.c]: `zalloc` expects to receive a `size_t`. So wouldn't it make more sense for `totalsize` to also be a `size_t`?

If so wouldn't it also make sense for `limbs` to be a size_t too? I realise we have to cast it to an `int` by the time we allocate it to `dsize`....but we're only doing that for historical reasons. I'd rather not propagate the historical reasons further than we have to.

File: crypto/fn/fn_lib.c [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+    /*
+     * If the size ends up being smaller than the bookkeeping struct,
+     * we know that the size calculation has wrapped around and is
+     * therefore invalid.  Also, we don't allow zero byte numbers.
+     */
+    if (totalsize <= (int)sizeof(OSSL_FN))
```
Comment (mattcaswell) [inline crypto/fn/fn_lib.c]: If `totalsize` was a size_t we wouldn't need this cast.

File: crypto/fn/fn_lib.c [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    return ret;
+}
+
+static void ossl_fn_free_internal(OSSL_FN *f, bool clear)
+{
+    if (f == NULL)
+        return;
+
+    int limbssize = f->dsize * sizeof(OSSL_FN_ULONG);
+    int totalsize = limbssize + sizeof(OSSL_FN);
```
Comment (mattcaswell) [inline crypto/fn/fn_lib.c]: Again - it makes more sense to me for `limbsize` and `totalsize` to be `size_t` because that's what we pass  to `OPENSSL_clear_free` and `OPENSSL_cleanse`

File: crypto/fn/fn_lib.c [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+{
+    return (size + 7) / 8;
+}
+
+OSSL_FN *OSSL_FN_new_bits(int size)
+{
+    return OSSL_FN_new_bytes(bits_to_bytes(size));
+}
+
+OSSL_FN *OSSL_FN_secure_new_bits(int size)
```
Comment (mattcaswell) [inline crypto/fn/fn_lib.c]: Again, I would make all of these sizes `size_t`

File: crypto/fn/fn_local.h [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+#  define OSSL_FN_DEC_FMT1      "%lu"
+#  define OSSL_FN_DEC_FMT2      "%019lu"
+# endif
+
+/*
+ * 64-bit processor other than LP64 ABI
+ */
+# ifdef SIXTY_FOUR_BIT
+#  undef BN_LLONG
+#  undef BN_ULLONG
```
Comment (mattcaswell) [inline crypto/fn/fn_local.h]: Not sure I understand this. Why are we undefing BN related things?

File: include/crypto/fn.h
Comment (mattcaswell) [inline include/crypto/fn.h:1]: There are some BIGNUM things in libssl. So, I wonder whether this will eventually need to move to include/internal/fn.h?

File: test/fn_internal_test.c [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    int ret = 1;
+    OSSL_FN *f = NULL;
+
+    /*
+     * Note that our uses of TEST_uint_ and TEST_size_t_ functions don't
+     * always correspond exactly to the types of the arguments.  However,
+     * the normal type coersion done in C allows this to be done without
+     * friction.
+     */
+    if (!TEST_ptr(f = OSSL_FN_new_bits(sizeof(OSSL_FN_ULONG) * 16 - 17))
```
Comment (mattcaswell) [inline test/fn_internal_test.c:51]: `sizeof(OSSL_FN_ULONG) * 16 - 17` seems quite arbitrary and strange. It took me some time to convince myself that the correct outcome is always 2 for `dsize`. Some commentary might be helpful.

Also, should we test out different boundary conditions and make sure we get the correct dsize output?

Shouldn't we also test the `new_bytes` and `new_limbs` variants too?

Comment (levitte) [reply test/fn_internal_test.c:51]: > Shouldn't we also test the `new_bytes` and `new_limbs` variants too?

We *are*...  indirectly.

File: crypto/fn/fn_local.h (hunk 12 above)
Comment (levitte) [reply crypto/fn/fn_local.h]: Ah, I forgot to rename those

File: include/crypto/fn.h
Comment (levitte) [reply include/crypto/fn.h:1]: Future PR?

Comment (levitte) [reply include/crypto/fn.h:1]: Actually....  the BIGNUMs there are only used for transport of data, yeah?  At least within libssl?  'cause libssl itself doesn't *operate* on them, yeah?

I'm not sure I'd bother very much with those, at least on a first and second pass, 'cause I assume that anything that's done with them is going to go through BN_ functions, which are supposed to become wrappers around OSSL_FN_ functionality.

**Comment by levitte** [comment]
Regarding the input to allocators, I just did change them to take a `size_t` size instead of an `int` size, and I'm again surprised that my compiler doesn't complain when I assign a `size_t` to an `int`.

The trouble I was seeing is a future plethora of `(int)` casts, especially in the interface between BN_ functions and OSSL_FN_ functions, since the BN_ code it littered with `int`s...  but, I just experimented a little, and there seems to be no need, at least on my laptop?
(of course, I'm sure I suffer from the "works on my machine" syndrome, and will get bitten sooner or later)

File: test/fn_internal_test.c (hunk 13 above)
Comment (levitte) [reply test/fn_internal_test.c:51]: I've made commentary related to both questions

File: include/crypto/fn.h
Comment (levitte) [reply include/crypto/fn.h:1]: But either way, I'm leaving that for later

**Comment by levitte** [comment]
The deed is done, I've converted `int` to `size_t` for all sizes where I could.  The sizes in the OSSL_FN struct itself remain `int`, though

**Comment by levitte** [comment]
>  (of course, I'm sure I suffer from the "works on my machine" syndrome, and will get bitten sooner or later)

Aha yup, Windows bites

File: include/crypto/fn.h
Comment (mattcaswell) [reply include/crypto/fn.h:1]: Fair enough

File: include/crypto/fn.h [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+# ifdef  __cplusplus
+extern "C" {
+# endif
+
+/*
+ * 64-bit processor with LP64 ABI
+ */
+# ifdef SIXTY_FOUR_BIT_LONG
+#  define OSSL_FN_ULONG         unsigned long
```
Comment (mattcaswell) [inline include/crypto/fn.h]: I guess the answer is going to be "because that's what we do in the BIGNUM" code - but I'll ask anyway. Why do we use define here (and other places) instead of typedef?

Comment (levitte) [reply include/crypto/fn.h]: To answer that question, I went looking to see if we're depending on BN_ULONG being a macro:

``` console
$ git grep '# *if.*BN_ULONG'
include/internal/constant_time.h:#ifdef BN_ULONG
```

The question is if that's enough of an argument.

I'm not really sure why there's that check in that header file, 'cause I can't imagine any reason why `BN_ULONG` wouldn't be defined.

Comment (levitte) [reply include/crypto/fn.h]: You know...  I think I'm gonna make OSSL_FN_ULONG a typedef.  I really see no reason why not, now that you put the light on it

Comment (levitte) [reply include/crypto/fn.h]: While I'm at it, I don't really see a need to avoid `uint64_t` and `uint32_t`, and leave it to the system to choose the right underlying type.

Comment (levitte) [reply include/crypto/fn.h]: Done

Comment (levitte) [reply include/crypto/fn.h]: Actually....  because OSSL_FN_ULONG should be compatible with BN_ULONG, and the latter is public, we're going to have OSSL_FN_ULONG be equivalent to BN_ULONG.  I'm changing things to make that explicit in a PR to come, but might as well have that be part of this PR...

Comment (levitte) [reply include/crypto/fn.h]: Done

File: crypto/fn/fn_local.h [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+#  define OSSL_FN_BITS4         16
+#  define OSSL_FN_MASK2         (0xffffffffL)
+#  define OSSL_FN_MASK2l        (0xffff)
+#  define OSSL_FN_MASK2h1       (0xffff8000L)
+#  define OSSL_FN_MASK2h        (0xffff0000L)
+#  define OSSL_FN_DEC_CONV      (1000000000L)
+#  define OSSL_FN_DEC_NUM       9
+#  define OSSL_FN_DEC_FMT1      "%u"
+#  define OSSL_FN_DEC_FMT2      "%09u"
+# endif
```
Comment (levitte) [inline crypto/fn/fn_local.h]: I'm dropping this cruft.  It's copied from `crypto/bn/bn_local.h`, but is currently unused, so I'm thinking that it might be brought in piece-meal as the need arises.

File: include/openssl/bn_limbs.h [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+/*
+ * 64-bit processor other than LP64 ABI
+ */
+# ifdef SIXTY_FOUR_BIT
+#  define BN_ULONG        unsigned long long
+#  define BN_BYTES        8
+# endif
+
+# ifdef THIRTY_TWO_BIT
+#  define BN_ULONG        unsigned int
```
Comment (mattcaswell) [inline include/openssl/bn_limbs.h:44]: I do wonder why we jump through all these type hoops. These days we require `uint64_t` to be present on all platforms. Why not just make `BN_ULONG` an alias for it?

Comment (levitte) [reply include/openssl/bn_limbs.h:44]: We adapt to the expected (arithmetic) register size.  Among others, our assembler stuff depends on it.

We could change things to use `uint32_t` and `uint64_t`.  However, I'm not ready to make that change in this PR.  Considering all the hackery there is in BIGNUM code, I prefer doing that in another PR and not overload this one, yeah?

**Comment by levitte** [comment]
After a small conversation with @mattcaswell, we decided to drop
`is_negative` and `is_acquired` for now.

It's not very probably that we'll need to support negative numbers in
`OSSL_FN`, and "acquiring" the `OSSL_FN` of a `BIGNUM` isn't
implemented at all yet, and might never be needed, so we chose to
defer those details until they are actually needed.

**Comment by levitte** [comment]
@mattcaswell, does your approval stand?

**Comment by levitte** [comment]
The MacOS failure isn't relevant

**Review by mattcaswell — APPROVED** [review]
LGTM

**Comment by mattcaswell** [comment]
Ping @t8m

**Comment by levitte** [comment]
Merged into feature/ossl_fn:

95fad8bee3cb88e947ab752a9ff1e159035fd243 Add the fixed number type (OSSL_FN) and its allocators and deallocator
